### ISSUE TITLE
Incorrectly callibrated screen in older devices (Kobo Mini, Kobo Touch)

### ISSUE THREAD

**Issue by nicolascolla** [koreader/koreader#10642]
The touch screen while using older devices (in my case Kobo Mini, Kobo Touch N905) is poorly callibrated, so much so that I have not been able to read any books.

This is a problem that seems to go at least as far back as koreader 2022.08 (the oldest version I have tried). This is a shame since the Kobo Mini has a very good form factor but nickel is very slow.

The upper part of the screen can't be used, tapping the top results in a tap in the middle of the screen, and tapping the bottom results in a tap in the middle of the screen, too. This happens both when using Kobo Start Menu or loading Koreader from nickel with kfmon.

**Comment by NiLuJe** [comment]
The Mini has (to my knowledge, as I've never had one) no known input quirks (besides a bezel that might impede accuracy near the edges), and we've never heard anything about it misbehaving, ever.

The touch is another kettle of fish, and you haven't given us enough details to go on, but it might lead to an obvious culprit: #10107 

TL;DR: Way more details and verbose debug logs, please.

**Comment by NiLuJe** [comment]
> TL;DR: Way more details and verbose debug logs, please.

On a recent FW versions, and *without* KSM, as it is quite likely guaranteed to break something here.

**Comment by NiLuJe** [comment]
> On a recent FW version

That may involve some hoop-jumping on the Mini, as some releases skipped the kernel binaries.

I don't recall the details, but it'll quite likely going through the latest "official" release for the Mini, the 3.19 re-release.
It should then be safe to jump to a recent kobo4 package from there.

**Comment by NiLuJe** [comment]
> (besides a bezel that might impede accuracy near the edges)

On that note, because I thought I'd murdered that hack long ago, check if removing the viewport entirely helps?

https://github.com/koreader/koreader/blob/346d8eb83e00e491e5ebe58f187130d5e9fec1aa/frontend/device/kobo/device.lua#L179

**Comment by nicolascolla** [comment]
@NiLuJe on my first round of testing with the Kobo Mini:

-Removing the viewport didn't help
-Uninstalling KSM didn't help
-Upgrading to 3.19.5761-s, then installing koreader with your package (https://www.mobileread.com/forums/showthread.php?t=314220) didn't help either.

This is very weird since a Kobo Glo HD with a firmware 3.14ish and KSM works like a charm. But both the Mini and the touch N905C exhibit this behaviour.

I attach crash.log. Please tell me if you'd like to peek at anything else.

[crash.log](https://github.com/koreader/koreader/files/11951094/crash.log)

**Comment by nicolascolla** [comment]
Kobo Touch N905C with latest firmware, installed via https://www.mobileread.com/forums/showthread.php?t=314220 and launched via NickelMenu: Works like a charm.

So it seems to be a problem with older firmwares.

**Comment by nicolascolla** [comment]
Probably off-topic, but 

1. Do you know of a way to get a newer firmware running in the Mini?

and

2. Is there a way to run koreader directly on a modern firmware, without going through nickel first?

**Comment by NiLuJe** [comment]
> This is very weird since a Kobo Glo HD with a firmware 3.14ish and KSM works like a charm. But both the Mini and the touch N905C exhibit this behaviour. And the behaviour is identical in both models, so I'm wondering if it's not a device-specific problem but a problem with low-res screens?

The Glo HD is far newer, it's as simple as that.

Nearly *every* Kobo device handles input in slightly different ways, so, yes, it *is* device-specific ;).

> Please tell me if you'd like to peek at anything else or need more info.

It's not verbose debug, so the only thing I can infer is that it is indeed running on an old FW, which doesn't really help us ;).

> Probably off-topic, but is there a way to run koreader directly on a modern firmware, without going through nickel first? So I can nuke KSM and be happy ever after.

[No, precisely because bypassing nickel may lead to hard-to-debug shenanigans like yours](https://github.com/koreader/koreader/issues/9781).

**Comment by NiLuJe** [comment]
> anything else or need more info

I doubt it'll change much because it's old as hell, and a Mini to boot, but do try to move to 4.37 via the kobo4 package (it certainly can't hurt, despite Nickel not offering it).

Other than that, you can unpack this on your device and run the devcap script over SSH follow the directions, and post the generated logs, in case we somehow missed a Mini input quirks than nobody noticed in years?
(And I distinctly remember active contributors having access to the device at one point, so I'm doubtful ;)).
[Kobo-DevCap-Test.tar.gz](https://github.com/koreader/koreader/files/11951350/Kobo-DevCap-Test.tar.gz)

**Comment by nicolascolla** [comment]
Thank you very much for your help, after upgrading to the latest Kobo Glo firmware on the Mini, koreader works perfectly fine, so it is working as intended. I will close this.

**Comment by NiLuJe** [comment]
Well, that's... mildly weird, but I'll take it ;o).

**Comment by nicolascolla** [comment]
@NiLuJe, just out of curiosity I decided to try every version of Koreader on my Mini, firmware 3.19.x with KSM until I found the last version that works. 2022.07 works like a charm, while 2022.08 already has this problem. So this "regression" occured somewhere between 2022.07 and 2022.08. I think the culprit might be https://github.com/koreader/koreader/commit/2678d2c49b375eb20b95161e0ce8b689d648ece4, but I have no way of testing this.

As I said before, koreader works perfectly in older Kobos running newer firmwares and NickelMenu. And I'm quite happy sticking with 2022.07 until the end of time. Just thought you guys would find it interesting to know.

**Comment by NiLuJe** [comment]
The Mini is "interesting" because of its peculiar EoL status: it should *guarantee* that there's no kernel update for it, ever.

As such, the input protocol *cannot* be different, so this still doesn't make much sense ;o).

(But, yes, it is one of few not-MT-capable devices, so this commit may not be entirely irrelevant).

What's more likely to have happened, though, is that Nickel sets the device up in a different framebuffer state on newer FWs, and that's what we expect it to be for proper rotation handling.

Can't say much more without access to the device (or at least a devcap run from Nickel on both of those FWs).

**Comment by NiLuJe** [comment]
And looking at the full [diff](https://github.com/koreader/koreader/compare/v2022.07...v2022.08), there's a crapload of input reworks in there ;D.
